library/macos/FLETextInputPlugin.m

```diff
@@ -238,13 +245,11 @@ - (void)doCommandBySelector:(SEL)selector {
 }
 
 - (void)insertNewline:(id)sender {
-  // This method is called when the user hits the Enter key. Since the embedder can't distinguish
-  // if the current widget is multiline or not, this method will be considered a "submit" action,
-  // To insert a new line, the user should hit ctrl + Enter to call insertLineBreak.
-  // There is a PR in Flutter to identify if the widget is multiline, and act accordingly
-  // (https://github.com/flutter/flutter/pull/23015). Once merged, this action should be changed
-  // to kNewLine.
-  [_channel invokeMethod:kPerformAction arguments:@[ _activeClientID, kDone ]];
+  if ([self.activeModel.inputType isEqualToString:kMultilineInputType]) {
+    [self insertText:@"\n" replacementRange:self.activeModel.selectedRange];
+  }
+  [_channel invokeMethod:kPerformAction
+               arguments:@[ _activeClientID, self.activeModel.inputAction ]];
```

Comment:
Note that if you make inputAction nullable, this needs special handling because putting nil into an array is a runtime exception.

Reply:
Made them nonnull, so no special handling needed.

Reply:
Re-opening, since currently this chain is not actually nonnull, so may or may not need changes depending on what happens with the other comment.

Reply:
Made the init method nullable, and raised an error if the creation of the model returns nil at line 123.